devops/pypi-release.yml

```diff
@@ -0,0 +1,73 @@
+# Pipeline Definition for Releases to PyPI
+
+# This Azure DevOps pipeline is for releasing fairlearn to PyPI. It performs the
+# following:
+# 1. Predeployment validation similar to the regular builds
+# 2. Release to PyPI-Test
+#    - Create package
+#    - Upload
+#    - Run tests against package on PyPI-Test
+# 3. Release to PyPI itself, repeating the same stages
+
+# Note that the pipeline MUST be queued with a variable RC_VERSION set to an
+# integer value. This is used to make the packages on PyPI-Test 'release
+# candidates' so that minor fixes can be made without requiring a separate checkin
+# to release to PyPI itself. See the logic in __init__.py
+
+# TODO:
+# Use Azure KeyVault so that uploads to PyPI actually occur
+
+variables:
+  poolImage: "ubuntu-latest"
+  poolPythonVersion: 3.6
+
+trigger: none # No CI build
+
+pr: none # Not for pull requests
+
+
+stages:
+- stage: PreDeploymentValidation
+  pool:
+    vmImage: $(PoolImage)
+    
+  jobs:
+  - template: all-tests-job-template.yml
+    parameters:
+      name: Linux
+      vmImage: 'ubuntu-16.04'
+      pyVersions: [3.5, 3.6, 3.7] 
+
+  - template: all-tests-job-template.yml
+    parameters:
+      name: Windows
+      vmImage:  'vs2017-win2016'
+      pyVersions: [3.5, 3.6, 3.7]
+      
+  - template: all-tests-job-template.yml
+    parameters:
+      name: MacOS
+      vmImage:  'macOS-10.13'
+      pyVersions: [3.6, 3.7] 
+
+  - template: notebook-job-template.yml
+    parameters:
+      pyVersions: [3.6]
+
+
+# =================================================================================================================
+
+# We rely on the implicit dependence on the previous stage
```

Comment:
Suggest using a stage with explicit dependence on success of prior stage.

Reply:
I had thought about that. The issue is that since I generate things out of templates, the names for the `dependsOn` clauses would have to be duplicated as well. I felt it was better to leave everything as implicitly dependent rather than duplicate the name generation logic. I'm persuadable on that point, though.